### DISCUSSION TITLE
sql server merge replication

### DISCUSSION THREAD

**Discussion by harishmk** [keycloak/keycloak#32871] [Q&A]
Hi All,

I am planning to deploy a Keycloak cluster across multiple datacenters, using SQL Server as the database. I am considering setting up [merge replication](https://learn.microsoft.com/en-us/sql/relational-databases/replication/merge/merge-replication) so that each datacenter will have a local copy of the database.

Merge replication adds a column and a unique index to every table. Will this setup be compatible with Keycloak?

Merge replication can encounter issues if each node performs heavy insertions or updates, leading to potential merge conflicts. How heavily does Keycloak modify the database? We are not using the persistent-user-sessions feature. Does Keycloak perform insertions or updates when a user logs in or generates tokens? Is it accurate to say that all database updates are triggered by operations in the admin console?

thanks.

**Comment by harishmk** [comment]
As a test, we set up two databases with merge replication between them. All the tables now have an additional column and an extra unique index.

On the face of it Keycloak appears to be functioning correctly with this setup, obviously our testing was very limited.
However, we are unsure if this configuration might affect other features or future version/schema upgrades.

Another aspect to consider is the frequency of database updates. If all cluster nodes write to the database on every login, session creation, token generation, etc., there could be a higher chance of merge conflicts. However, if database writes occur only through the admin interface, we can mitigate this risk by limiting the admin interface to a single node in the cluster, thereby reducing the likelihood of merge conflicts.

**Reply by willibrandon** [reply]
Merge replication uses a default priority-based conflict resolver and the behavior depends on the subscription type. For client type subscriptions, the default behavior in the event of a conflict is the first change written to the Publisher wins (first-to-publisher-wins). Since one change wins, that means the other change loses, which means a potential for data loss 😞 Keep a close eye out for and set an alert for any conflicts that occur.
